Sources/NIO/ChannelOption.swift

```diff
@@ -327,3 +327,39 @@ extension ChannelOptions {
         }
     }
 }
+
+// Provides convenience methods for `ChannelOption.socket`
+extension ChannelOptions {
+    
+    public struct Socket {
+        
+        /// Turn on debugging info recording
+        public static let enableDebugging = ChannelOptions.Types.SocketOption(level: SOL_SOCKET, name: SO_DEBUG)
```

Comment:
I'm happy with this, but this way, users still need to specify `1` as 'true' instead of `true`. We _could_ also define a new type in `ChannelOptions.Types` which abstracts that away. @Lukasa WYDT?

    ...Options(ChannelOptions.Socket.enableDebugging, value: 1)

vs

    ...Options(ChannelOptions.Socket.enableDebugging, value: true)

at the expense of more code in this PR. I favour the latter but I'm ok going with the first. If we were to choose the second, then we could make the `ChannelOptions.Types.BSDSocketOption<ValueType>`.

Reply:
How does that interact with the pre-existing `SocketOptionProvider` protocol?

Reply:
I'm ok with either, but would lean toward putting in a separate PR...

Reply:
Scratch that, this is a super-simple change, so I will add in this PR

Reply:
@weissi I've done quite a bit of experimenting for the second option, the code changes are significant so I think this is best placed in another PR that I can work on. What do you think?

Reply:
@Lukasa I don't think we need the more complicated `SocketOptionProvider` protocol. Here's what I mean with the example of just a replacement for `SO_REUSEADDR`:

```
diff --git a/Sources/NIO/BaseSocketChannel.swift b/Sources/NIO/BaseSocketChannel.swift
index 7bbd5ac9..0df8b2ce 100644
--- a/Sources/NIO/BaseSocketChannel.swift
+++ b/Sources/NIO/BaseSocketChannel.swift
@@ -482,6 +482,8 @@ class BaseSocketChannel<SocketType: BaseSocketProtocol>: SelectableChannel, Chan
         switch option {
         case let option as ChannelOptions.Types.SocketOption:
             try self.setSocketOption0(level: option.level, name: option.name, value: value)
+        case let option as ChannelOptions.Types.IntAsBoolSocketOption:
+            try self.setSocketOption0(level: option.level, name: option.name, value: value as! Bool ? 1: 0)
         case _ as ChannelOptions.Types.AllocatorOption:
             bufferAllocator = value as! ByteBufferAllocator
         case _ as ChannelOptions.Types.RecvAllocatorOption:
diff --git a/Sources/NIO/ChannelOption.swift b/Sources/NIO/ChannelOption.swift
index 483d1f3d..fb5af6c6 100644
--- a/Sources/NIO/ChannelOption.swift
+++ b/Sources/NIO/ChannelOption.swift
@@ -66,6 +66,23 @@ public typealias AllowRemoteHalfClosureOption = ChannelOptions.Types.AllowRemote
 extension ChannelOptions {
     public enum Types {
 
+        public struct IntAsBoolSocketOption: ChannelOption, Equatable {
+            public typealias Value = Bool
+
+            public var level: SocketOptionLevel
+            public var name: SocketOptionName
+
+            /// Create a new `SocketOption`.
+            ///
+            /// - parameters:
+            ///     - level: The level for the option as defined in `man setsockopt`, e.g. SO_SOCKET.
+            ///     - name: The name of the option as defined in `man setsockopt`, e.g. `SO_REUSEADDR`.
+            public init(level: SocketOptionLevel, name: SocketOptionName) {
+                self.level = level
+                self.name = name
+            }
+        }
+
         /// `SocketOption` allows users to specify configuration settings that are directly applied to the underlying socket file descriptor.
         ///
         /// Valid options are typically found in the various man pages like `man 4 tcp`.
@@ -230,6 +247,8 @@ public struct ChannelOptions {
         .init(level: level, name: name)
     }
 
+    public static let allowLocalAddressReUse = Types.IntAsBoolSocketOption(level: .init(SOL_SOCKET), name: SO_REUSEADDR)
+
     /// - seealso: `AllocatorOption`.
     public static let allocator = Types.AllocatorOption()
 
diff --git a/Sources/NIOChatClient/main.swift b/Sources/NIOChatClient/main.swift
index b03af9ea..26135af2 100644
--- a/Sources/NIOChatClient/main.swift
+++ b/Sources/NIOChatClient/main.swift
@@ -44,7 +44,7 @@ private final class ChatHandler: ChannelInboundHandler {
 let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
 let bootstrap = ClientBootstrap(group: group)
     // Enable SO_REUSEADDR.
-    .channelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+    .channelOption(ChannelOptions.allowLocalAddressReUse, value: true)
     .channelInitializer { channel in
         channel.pipeline.addHandler(ChatHandler())
     }
diff --git a/Sources/NIOChatServer/main.swift b/Sources/NIOChatServer/main.swift
index 9924723f..cc0e6769 100644
--- a/Sources/NIOChatServer/main.swift
+++ b/Sources/NIOChatServer/main.swift
@@ -121,7 +121,7 @@ let group = MultiThreadedEventLoopGroup(numberOfThreads: System.coreCount)
 let bootstrap = ServerBootstrap(group: group)
     // Specify backlog and enable SO_REUSEADDR for the server itself
     .serverChannelOption(ChannelOptions.backlog, value: 256)
-    .serverChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+    .serverChannelOption(ChannelOptions.allowLocalAddressReUse, value: true)
 
     // Set the handlers that are applied to the accepted Channels
     .childChannelInitializer { channel in
```

@Davidde94 is this really much more code?

Reply:
for all the actually boolean ones, we need only one type (`BoolAsIntSocketOption`) and then we can go ahead and simply define

```swift
public static let allowLocalAddressReUse = Types.IntAsBoolSocketOption(level: .init(SOL_SOCKET), name: SO_REUSEADDR)
public static let allowPortReUse = Types.IntAsBoolSocketOption(level: .init(SOL_SOCKET), name: SO_REUSEPORT)
public static let enableDebugging = Types.IntAsBoolSocketOption(level: .init(SOL_SOCKET), name: SO_DEBUG)
...
```

Reply:
@weissi I was responding only to the proposal to have `ChannelOptions.Types.BSDSocketOption<ValueType>`. Special-casing bool isn't a big deal to my eye, but widening the scope raises questions about whether we're killing `SocketOptionProvider` or not.

Reply:
@Lukasa ah sorry, at first I thought we might want to use one type (`BSDSocketOption<ValueType>`) instead of specialised ones (`BoolAsIntSocketOption`) because I thought it's easier in the code but it's not because we still have to match each type separately (`BSDSocketOption<Bool>`, `BSDSocketOption<...>`). That's why I then quickly switched to `BoolAsIntSocketOption` (terrible name...).

Regarding `SocketOptionProvider`: I thought we need that if someone wants to set an option for which we don't have special support and that is wider than a `CInt`?

Reply:
We do, but once we go down the road of `BSDSocketOption<Value>` there is precious little reason not to allow `Value` to be essentially any type and just bitwise-copy it, which would cover the use-case for `SocketOptionProvider` as well.

Reply:
@Lukasa aah I see. I don't _think_ it would work with a truly generic `BSDSocketOption<Value>` because we need to match the type in this switch

```
        switch option {
        case let option as ChannelOptions.Types.SocketOption:
            try self.setSocketOption0(level: option.level, name: option.name, value: value)
        case let option as ChannelOptions.Types.IntAsBoolSocketOption:
            try self.setSocketOption0(level: option.level, name: option.name, value: value as! Bool ? 1: 0)
```

and we can't just write

```
    case let option as ChannelOptions.Types.BSDSocketOption<T>:
```

we need a concrete `T` here and can't just match all of them. We could match `BSDSocketOption<Bool>`, and `BSDSocketOption<CInt>`, and `BSDSocketOption<linger>`, ... but that way we wouldn't support arbitrary wide stuff. Am I missing anything?

What we _could_ probably do is something like

```
        public struct GenericBSDSocketOption: ChannelOption, Equatable {
            public typealias Value = [UInt8] // byte representation of the value

            public var level: SocketOptionLevel
            public var name: SocketOptionName

            public init(level: SocketOptionLevel, name: SocketOptionName) {
                self.level = level
                self.name = name
            }
        }
```

and then value would just be the bytes (bitwise-copied) for the value to set.

Reply:
and we'd set the value by doing

```
var crazyOptionValue: linger = linger(...)
.channelOption(GenericBSDSocketOption(level: SOL_SOCKET, name: SO_LINGER), value: withUnsafeBytes(of: &crazyOptionValue, Array.init))
```